storage/merge.go

```diff
@@ -487,34 +490,47 @@ func (c *chainSampleIterator) Next() bool {
 		}
 	}
 
+	if c.curr == nil {
+		return false
+	}
+
+	var currt int64
 	for {
-		if c.curr != nil && c.curr.Next() {
-			currt, _ := c.curr.At()
+		currt = math.MaxInt64
+		if c.curr.Next() {
+			currt, _ = c.curr.At()
 			if currt == c.lastt {
 				// Ignoring sample for the same timestamp.
 				continue
 			}
 			if len(c.h) == 0 {
-				return true
-			}
-			nextt, _ := c.h[0].At()
-			if currt < nextt {
-				c.lastt = currt
-				return true
+				// curr is the only iterator remaining
+				// No need to check with the heap.
+				break
 			}
-			heap.Push(&c.h, c.curr)
 		} else if len(c.h) == 0 {
+			// No iterator left to iterate.
 			c.curr = nil
 			return false
 		}
 
+		// Check current iterator with the top of the heap.
+		if nextt, _ := c.h[0].At(); currt < nextt {
+			// Current iterator has smaller timestamp than the heap.
+			break
+		}
+
+		// Pushing curr iterator back to heap to get the smallest timestamp.
+		heap.Push(&c.h, c.curr)
```

Comment:
Hm, this is wrong if c.curr.Next() returned false. I guess the previous one was more apt.